### PR TITLE
Use synthetic functions of kotlin jvm inline class

### PR DESCRIPTION
**Pull request by wplong11**
## Background

To support kotlin value class,
Synthetic Kotlin constructors should be available

revert: #1005

### CODE REVIEW

**Comment by pjfanning** [comment]
That code has been in jackson for years. Many users have used it knowingly or unknowingly. We can't just revert it because 1 user doesn't like it. No use cases have been provided. No description has been provided.

**Comment by wplong11** [comment]
@pjfanning Below is the use case. Before discussing the detailed design, I created a PoC.
https://github.com/FasterXML/jackson-module-kotlin/pull/693

**Comment by wplong11** [comment]
Since it's a Kotlin-related problem, I wonder if it's correct to support it in a Kotlin module...

> Many users have used it knowingly or unknowingly

But what you said worries me too. By the way, can you reproduce the issue in the original question? Deleting it now doesn't seem to be a problem. I hope you can help.

**Comment by k163377** [comment]
First, you can get a callable constructor without making this change.
Synthetic constructors have a pair of private, non-synthetic constructors that reside in a scope that can be parsed by `Jackson`.
The code is quite old, but the following may be helpful.
https://github.com/FasterXML/jackson-module-kotlin/compare/2.16...k163377:jackson-module-kotlin:github-413/get_constructor_backup
https://github.com/FasterXML/jackson-module-kotlin/issues/199#issuecomment-1013810769

This approach has the problem that annotations given to the constructor do not work, but `Kogera` has found a temporary solution.
https://github.com/ProjectMapK/jackson-module-kogera/blob/0ffefda68d5caa17b149d488e4e0054909e84319/src/main/kotlin/io/github/projectmapk/jackson/module/kogera/KotlinFeature.kt#L76

As long as there is a workaround, it does not seem so reasonable to change the treatment of composite constructors.

---

However, I too would like to target composite constructors for processing.

For example, what about adding a process to `AnnotationIntrospector` that would extract the synthetic constructors that I want to process from the class?
I am not familiar with the `databind` implementation, but I suspect such a feature would minimize the impact.

If this is also required in modules or libraries other than `Kotlin`, I think it would be worth considering implementation.

**Comment by kkurczewski** [comment]
Hi

As my PR was mentioned I will allow myself highlight problem which I have as well, assume I have class like this:
```
data class Something(val x: String, val y: String)
```
And I want to run code like this (it mimics Jackson ctor lookup):
```
Something::class.java.constructors.filter { !it.isSynthetic }.onEach { println("#1: $it") }.first()
```
And this will work. However, when you add an inline class to POJO it isn't gonna work:
```
data class SomethingWithInlineClass(val x: String, val y: kotlin.time.Duration)

SomethingWithInlineClass::class.java.constructors.filter { !it.isSynthetic }.onEach { println("2: $it") }.first()
```
Result is:
```
Exception in thread "main" java.util.NoSuchElementException: List is empty.
	at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:212)
	at com.fasterxml.jackson.module.kotlin.test.MainKt.main(main.kt:10)
	at com.fasterxml.jackson.module.kotlin.test.MainKt.main(main.kt)
```
In case of Jackson we get a more verbose message about missing creators. Sadly, there is no way to configure that behavior, there is no class equivalent like `NopAnnotationIntrospector`, `StdValueInstantiator` etc. that allow to customize that (there is a lot of `final` classes in chain) hence I don't have a place where I could use trick @k163377 propose.

**Comment by cowtowncoder** [comment]
Ok, I am worried too, although all unit tests do pass. But then again we don't have tests for auto-generated classes frameworks produce etc.
There is also no test showing this works (but I assume Kotlin module would have something).

I do think that improved handling should be added (but not via `AnnotationIntrospector`): ideally as part of full rewrite of property introspection. I still plan to tackle that before 2.16.0.
This probably wouldn't be added as part of rewrite BUT it:

1. could/should be tackled immediately after rewritten code works and
2. possibly should not be made before rewrite (as it might complicate)

So I won't be merging this before going on vacation, but it may be left open.

**Comment by cowtowncoder** [comment]
One more thing: allowing per-class inclusion of synthetic methods would be something I'd consider, but the question is how to access this without needing to sub-class is or have global configuration.
If code in question could access `AnnotationIntrospector`, that might be an option (although not necessarily great one). I may be forgetting some obvious choice, but pluggable "constructor classifier" could do.

The key, at any rate, would be to limit scope of changed visibility to Kotlin (-generated) classes; allow Kotlin module indicate changed rules.
If so, then other modules (often Scala module benefits from additions like this; but also some frameworks?) would benefit from this too, and risk would be limited.

**Comment by wplong11** [comment]
I will also check https://github.com/FasterXML/jackson-databind/pull/4066#issuecomment-1664351480

**Comment by wplong11** [comment]
@k163377 

>For example, what about adding a process to AnnotationIntrospector that would extract the synthetic constructors that I want to process from the class?
I am not familiar with the databind implementation, but I suspect such a feature would minimize the impact.

You mean you want to be able to set the constructor extraction strategy using Module interface (SimpleModule / KotlinModule)?

**Comment by kkurczewski** [comment]
@k163377 

> > `Something::class.java.constructors.filter { !it.isSynthetic }.onEach { println("#1: $it") }.first()`
> 
> I don't think I'm reading your argument well, but does not using `getDeclaredConstructors` have anything to do with the issue? https://docs.oracle.com/javase/jp/8/docs/api/java/lang/Class.html#getDeclaredConstructors--

Yes, my mistake - Jackson uses [declared constructors](https://github.com/FasterXML/jackson-databind/blob/2.16/src/main/java/com/fasterxml/jackson/databind/util/ClassUtil.java#L576). I had too many attempts and my memory starts to plays tricks on me. But I'm pretty sure my current impl still has problem with missing creator, but will take a second look on my PR and will try to use [this method](https://github.com/FasterXML/jackson-module-kotlin/compare/2.16...k163377:jackson-module-kotlin:github-413/get_constructor_backup#diff-e0d9bb1fc63d6a571a82ca6a54c0892b900cbcae40b3cd84c964c2875f7726a5R94).

**Comment by wplong11** [comment]
> since the instantiation of the value class requires calling box-impl after calling constructor-impl.

@k163377 Sorry, I didn't understand. 
The Value Class seems to be initialized correctly even if I called box-impl without calling constructor-impl.(https://github.com/FasterXML/jackson-module-kotlin/pull/694 - I added more test cases)
Is there any black magic I don't know about?

**Comment by k163377** [comment]
The `value class` may have a check process in its constructor.
If only `box-impl` is called, the check is actually skipped.
https://github.com/ProjectMapK/jackson-module-kogera/blob/81c6497eaa5609ede547567bc8ebdc961399e77d/src/test/kotlin/io/github/projectmapk/jackson/module/kogera/_integration/deser/value_class/NoSpecifiedTest.kt#L43-L48

**Comment by cowtowncoder** [comment]
Ok, this looks doable -- but I'll need to think about it a bit; I hope to be able to give better feedback (or approve) soon but have a big backlog to go through after vacation.

File: src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedCreatorCollector.java [hunk 1]
```diff
@@ -406,6 +411,6 @@ private final AnnotationMap collectAnnotations(AnnotatedElement main, AnnotatedE
 
     // for [databind#1005]: do not use or expose synthetic constructors
     private static boolean isIncludableConstructor(Constructor<?> c) {
```
[the remark this answers is not in the document]
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedCreatorCollector.java:412]: I don't think current `AnnotatedXxxCollector`s are optimal extension points, so suggested changes here are probably best we can do on short term. Before bigger rewrite,

**Comment by cowtowncoder** [comment]
Ok, I'd be happy to merge this, if we can change this to non-draft PR?

**Comment by k163377** [comment]
Personally, I am against merging this PR.

It contains at least some undesirable content from a `Kotlin` perspective.
It also looks like this could affect many projects that use `Kotlin`.
I don't think we should make changes with such a large impact when workarounds exist.

File: src/main/java/com/fasterxml/jackson/databind/ext/KotlinSupport.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+        Class<?> lastParam = params[params.length - 1];
+        return ctor.isSynthetic() && lastParam.getName().equals("kotlin.jvm.internal.DefaultConstructorMarker");
+    }
+
+    public static boolean isJvmInlineClassSyntheticBoxingFunction(Method method) {
+        return Modifier.isStatic(method.getModifiers())
+                && method.isSynthetic()
+                && method.getName().equals("box-impl");
+    }
```
Comment (k163377) [inline src/main/java/com/fasterxml/jackson/databind/ext/KotlinSupport.java:22]: As commented here, using only `box-impl` to deserialize `value class` may cause unexpected bugs.
https://github.com/FasterXML/jackson-databind/pull/4066#issuecomment-1665745100

Therefore, this change should be removed.

File: src/main/java/com/fasterxml/jackson/databind/ext/KotlinSupport.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+public class KotlinSupport {
+    public static boolean isJvmInlineClassSyntheticConstructor(Constructor<?> ctor) {
+        Class<?>[] params = ctor.getParameterTypes();
+        if (params.length == 0) {
+            return false;
+        }
+
+        Class<?> lastParam = params[params.length - 1];
+        return ctor.isSynthetic() && lastParam.getName().equals("kotlin.jvm.internal.DefaultConstructorMarker");
+    }
```
Comment (k163377) [inline src/main/java/com/fasterxml/jackson/databind/ext/KotlinSupport.java:16]: This code also covers constructors that do not actually contain `value class` as an argument.

The following is a class definition in `Kotlin` and its constructors taken from the decompiled result.
You can see that the `value class` is not used, but the argument of the synthetic constructor ends with `DefaultConstructorMarker`.

```kotlin
data class Data(val v: Int = 0)
```

```java
// decompiled

   public Data(int v) {
      this.v = v;
   }

   // $FF: synthetic method
   public Data(int var1, int var2, DefaultConstructorMarker var3) {
      if ((var2 & 1) != 0) {
         var1 = 0;
      }

      this(var1);
   }
```

I can't think of any other way to write it myself, but I think it should at least be supplemented with a comment.

**Comment by k163377** [comment]
I thought about it again, but it seems that simply including the synthetic constructor in the processing target does not solve the problem, since the deserialization of the `value class` requires two functions to be called in the correct order.
It seems to me that we need a function to make externally defined functions and processes available as `Creator`.

**Comment by cowtowncoder** [comment]
@k163377 I trust your judgment here -- I was under (wrong) impression that this alone would help solve the issue by not pruning certain synthetic constructors/factory methods. But if this alone won't help, yeah, no point in doing that.

That said, I am perfectly willing to add a little bit of code (like here) where necessary to support extensions such as Kotlin or Scala module, and have appropriate scope not to change behavior of types on "other" platforms.
But it sounds, unfortunately, that the code here isn't quite discriminative enough?

**Comment by cowtowncoder** [comment]
Looks like this PR isn't going anywhere and is out-of-date. Will close.

**Comment by JooHyukKim** [comment]
> Looks like this PR isn't going anywhere and is out-of-date. Will close.

WRT out-dated, stale issues, there is automated labeler as Github workflow for stale-ness we might be interested in?

**Comment by cowtowncoder** [comment]
@JooHyukKim thank you for suggestion!

My immediate reaction is that I am -1 for auto-close-issues Bots. Projects that use them seem to have very aggressive garbage collection, leading to (IMHO) hostile experience for submitters. So while I have sympathy for maintainers, personally I much prefer actual human beings closing issues over automated processes.

But this is only for Bots that actually close issues -- I would not be against auto-labeller?
As long as that has long enough period before adding label(s).

**Comment by JooHyukKim** [comment]
Np, @cowtowncoder 👍🏼 I agree, closing would be too agressive. And yes it will be just labeler. Okay, will get into it.

**Comment by cowtowncoder** [comment]
Sounds good -- and we could/should probably only enable it (if we do) for some of highest-traffic repos.

**Pull request closed without merging**